### PR TITLE
Support abort hooks (symmetrically to commit hooks) #77

### LINKED ISSUE #77: Support abort hooks (symmetrically to commit hooks)
`transaction` currently supports (before and after) commit hooks. It should symmetrically support abort hooks as well.

I need this feature in `dm.zodb.asynchronous`, more precisely in its `TransactionalScheduler`. This scheduler allows to schedule a function call at transaction commit; it uses a commit hook for this. If the transaction is aborted (instead of committed), a cleanup becomes necessary. This requires a (transaction) abort hook.

I implement this feature currently in `dm.transaction.aborthook` via monkey patching. If there is interest, I could implement it instead in `transaction`.

### PR DESCRIPTION
**Pull request by d-maurer**
Fixes #77

### CODE REVIEW

File: transaction/interfaces.py [hunk 1]
```diff
@@ -356,6 +356,74 @@ def getAfterCommitHooks():
...
+        names and values to be passed, or the default None (no keyword
+        arguments are passed).
+
+        Multiple hooks can be registered and will be called in the order they
+        were registered (first registered, first called).  This method can
+        also be called from a hook:  an executing hook can register more
+        hooks.  Applications should take care to avoid creating infinite loops
+        by recursively registering hooks.
+
+        Hooks are called only for a top-level abort. If the
```
Comment (jamadden) [inline transaction/interfaces.py]: I wonder if those are the best semantics. If someone wants to catch all possible aborts, they need to register two hooks: one for top-level aborts, and one for after commit, checking for `False`. It seems like it would be easier to also call the abort hooks from `_cleanup` (error handling during commit) so that all cases where a transaction is aborted can be handled with a single hook. Of course, that might raise the possibility of wanting to know whether the abort was explicit or implicit...

I'm not sure I have a strong opinion about this, but I figured it was worth discussing.

File: transaction/_transaction.py [hunk 2]
```diff
@@ -409,6 +416,67 @@ def _callAfterCommitHooks(self, status=True):
...
+            except:
+                # We need to catch the exceptions if we want all hooks
+                # to be called
+                self.log.error("Error in after abort hook exec in %s ",
+                               hook, exc_info=sys.exc_info())
+        # The transaction is already abortted. It must not have
+        # further effects after the abort.
+        for rm in self._resources:
+            try:
+                rm.abort(self)
```
Comment (jamadden) [inline transaction/_transaction.py]: It's not clear to me why we're calling `rm.abort` again in this loop?

File: transaction/_transaction.py [hunk 3]
```diff
@@ -409,6 +416,67 @@ def _callAfterCommitHooks(self, status=True):
...
+        # growing list is well-defined in Python.
+        for hook, args, kws in self._after_abort:
+            try:
+                hook(*args, **kws)
+            except:
+                # We need to catch the exceptions if we want all hooks
+                # to be called
+                self.log.error("Error in after abort hook exec in %s ",
+                               hook, exc_info=sys.exc_info())
+        # The transaction is already abortted. It must not have
```
Comment (jamadden) [inline transaction/_transaction.py]:
```suggestion
        # The transaction is already aborted. It must not have
```

File: transaction/_transaction.py [hunk 4]
```diff
@@ -409,6 +416,67 @@ def _callAfterCommitHooks(self, status=True):
...
+        if kws is None:
+            kws = {}
+        self._before_abort.append((hook, tuple(args), kws))
+
+    def _callBeforeAbortHooks(self):
+        # Call all hooks registered, allowing further registrations
+        # during processing.  Note that calls to addBeforeAbortHook() may
+        # add additional hooks while hooks are running, and iterating over a
+        # growing list is well-defined in Python.
+        for hook, args, kws in self._before_abort:
```
Comment (jamadden) [inline transaction/_transaction.py]: I wonder if calling the before abort hooks should be wrapped in `try/except:` handling.

I know the before *commit* hooks aren't wrapped in that, but that's ok: if they throw, it's possible to catch that and abort the transaction; one con do this to programatically implement dooming transactions, for example:
```python
try:
   tx.commit()
except DoomedTransaction:
  tx.abort()
```

So it's OK if a before-commit hook raises. But if a before abort hook raises, then we're left with no way to actually abort the transaction.

File: transaction/_transaction.py [hunk 5]
```diff
@@ -409,6 +416,67 @@ def _callAfterCommitHooks(self, status=True):
         self._after_commit = []
         self._before_commit = []
 
+    def getBeforeAbortHooks(self):
```
Comment (jamadden) [inline transaction/_transaction.py]: I feel like there's enough duplication across these four different hooks that some meta-programming to reduce duplication might be called for. That can be a separate PR though.

**Comment by d-maurer** [comment]
Jason Madden wrote at 2019-5-14 06:12 -0700:
>jamadden commented on this pull request.
>
>I left some questions that I'd like to gain some insight on. It might be good to get other's opinions too (/cc @tseaver @jimfulton @mgedmin)
>
>> @@ -120,6 +120,13 @@ def __init__(self, synchronizers=None, manager=None):
>         # List of (hook, args, kws) tuples added by addAfterCommitHook().
>         self._after_commit = []
>
>+
>
>Looks like an extra line of whitespace.

Possible. The repository `zopefoundation/transaction` has not
yet a "flake8" test (like other `zopefoundation` repositories).

> ...
>+        Hooks are called only for a top-level abort. If the
>
>I wonder if those are the best semantics. If someone wants to catch all possible aborts,
>they need to register two hooks: one for top-level aborts, and one for after commit, checking for `False`. It seems like it would be easier to also call the abort hooks from `_cleanup` (error handling during commit) so that all cases where a transaction is aborted can be handled with a single hook. Of course, that might raise the possibility of wanting to know whether the abort was explicit or implicit...

The code in the PR essentially comes from `dm.transaction.aborthook`.
There, I was surprised to find out that `transaction` makes the distinction
between an implicit abort (called internally during commit) and
an explicit (aka "top level") abort (I had not expected such a
distinction - and I wrote special documation to make this clear).
The implicit abort during commit already had its hook (the commit hook
with first parameter `False`) and I did not want to add
an additional kind of hooks for this case.

> ...
>It's not clear to me why we're calling `rm.abort` again in this loop
>(note: the loop after the after abort hooks run)?

I have copied the code from the "after commit hook call" handling.

The motivation behind seems to be: the hooks might have interacted
with the resource managers (sometimes also called data managers)
and left them in an unclean state. The "rm.abort()" calls
should ensure, that they become clean again.

> ...
>I wonder if calling the before abort hooks should be wrapped in `try/except:` handling.
> ...
>So it's OK if a before-commit hook raises. But if a before abort hook raises, then we're left with no way to actually abort the transaction.

You are right.

I will change this together with potential other changes we
decide upon.

> ...
>I feel like there's enough duplication across these four different hooks that some meta-programming to reduce duplication might be called for. That can be a separate PR though.

This is a matter of preferences.

I dislike duplication - potentially even more than you.
However, Jim disliked my implementation of `TransactionManager.run`
in terms of `TransactionManager.attempts` (a beautiful very
elegant implementation without duplication) in favor of one
in elementary terms (with quite some non-trivial logic duplicated
from `attempts`).
I concluded that the `transaction` "master"s prefer explicity
over conciseness and elegance.

**Comment by d-maurer** [comment]
The discussion in https://github.com/zopefoundation/transaction/issues/82 reminded my that all hooks must be cleared after commit/abort (which currently is not yet ensured).

**Comment by d-maurer** [comment]
`commit` likely calls `_free` and `_callAfterCommitHooks` in the wrong order: the `_free` informs the transaction manager that the current transaction has terminated; if the manager is in "implicit" mode, then it will (effectively) start a new transaction. The `_callAfterCommitHooks` has special code to clean up the resource managers in the case that a hook call has changed one of them. However, it looks at its own resource managers (cleared by `_free`) while the hook has potentially affected those of a new transaction.

File: transaction/interfaces.py (hunk 1 above)
Comment (d-maurer) [reply transaction/interfaces.py]: @jamadden 
> I wonder if those are the best semantics. If someone wants to catch all possible aborts, they need to register two hooks: one for top-level aborts, and one for after commit, checking for `False`....

I am not yet sure - but maybe, the `abort` hook would be sufficient: when I remember right, then a transaction with commit problems enters a "DOOMED" state and must be explicitly "aborted". Should this be true, then we could expect, that typically a top level `abort` follows a commit with problems.

Comment (d-maurer) [reply transaction/interfaces.py]: I have verified: in case of a commit problem, the transaction enters the `COMMITFAILED` state and becomes virtually unusable. The natural thing to do in this case is to `abort` the transaction. in order to get a new (clean) one. I have changed the documentation accordingly.

File: transaction/_transaction.py (hunk 3 above)
Comment (d-maurer) [reply transaction/_transaction.py]: I have put the call hook logic into a centralized function (as you have recommended). As a consequence, the above comment disappeared.

File: transaction/_transaction.py (hunk 4 above)
Comment (d-maurer) [reply transaction/_transaction.py]: Changed as suggested.

**Comment by d-maurer** [comment]
@jamadden I have followed most of your suggestions. Would you like to take another look?
